fw/ringbuffer.c

```diff
@@ -0,0 +1,248 @@
+	proc_file_is_open = true;
+	return 0;
+}
+
+int
+tfw_ringbuffer_file_close(struct inode *ino, struct file *filp)
+{
+	int cpu;
+	proc_file_is_open = false;
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+		atomic_set(&rba->unmapped, 1);
+	}
+	return 0;
+}
+
+int
+tfw_ringbuffer_file_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+	TfwRingbufferArea *rba, *this_rba = this_cpu_ptr(&rb_area);
+	unsigned long pfn, size, area_size, area_pages;
+	int cpu;
+
+	area_size = RBA_AREA_SIZE(this_rba);
+	size = vma->vm_end - vma->vm_start;
+	if (size > area_size)
+		return -EINVAL;
+
+	area_pages = area_size / PAGE_SIZE;
+
+	cpu = vma->vm_pgoff / area_pages;
+	if (cpu >= num_online_cpus())
+		return -EINVAL;
+
+	rba = per_cpu_ptr(&rb_area, cpu);
+	pfn = page_to_pfn(rba->pg);
+
+	if (remap_pfn_range(vma, vma->vm_start, pfn, size, vma->vm_page_prot))
+		return -EAGAIN;
+
+	if (size == area_size)
+		atomic_set(&rba->unmapped, 0);
+
+	return 0;
+}
+
+void
+tfw_ringbuffer_test_set_unmapped(int unmapped)
+{
+	int cpu;
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+		atomic_set(&rba->unmapped, unmapped);
+	}
+}
+
+int
+tfw_ringbuffer_test_read(char *data, u32 size)
+{
+	u32 head, tail;
+	int cpu;
+
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+		TfwRingbuffer *rb = rba->rb;
+		u32 size_for_cpu;
+
+		head = smp_load_acquire(&rb->head);
+		tail = rb->tail;
+
+		size_for_cpu = min(size, head - tail);
+		size -= size_for_cpu;
+
+		while (size_for_cpu) {
+			u32 mtail, cur_size;
+
+			mtail = tail & rb->mask;
+
+			cur_size = min(size_for_cpu, rb->size - mtail);
+			memcpy_fast(data, rb->data + mtail, size_for_cpu);
+			data += cur_size;
+			tail += cur_size;
+			size_for_cpu -= cur_size;
+		}
+
+		smp_store_release(&rb->tail, tail);
+		smp_mb();
+	}
+
+	return size;
+}
+
+int
+tfw_ringbuffer_init(unsigned int size)
+{
+	int cpu, res;
+	unsigned int order;
+
+	if (size < TFW_RINGBUFFER_MIN_SIZE
+		|| size > TFW_RINGBUFFER_MAX_SIZE
+		|| !is_power_of_2(size))
+		return -EINVAL;
+
+	order = get_order(AREA_SIZE(size));
+
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+
+		rba->pg = alloc_pages_node(cpu_to_node(cpu), GFP_KERNEL, order);
+		if (rba->pg == NULL) {
+			res = -ENOMEM;
+			goto err;
+		}
+
+		rba->rb = (TfwRingbuffer *)page_address(rba->pg);
+
+		rba->rb->size = size;
+		rba->rb->mask = size - 1;
+		rba->rb->head = 0;
+		rba->rb->tail = 0;
+		rba->rb->cpu = cpu;
+		rba->rb->max_cpu_num = num_online_cpus();
+
+		atomic_set(&rba->unmapped, 1);
+	}
+
+	return 0;
+
+err:
+	tfw_ringbuffer_cleanup();
+
+	return res;
+}
+
+void
+tfw_ringbuffer_cleanup(void)
+{
+	int cpu;
+
+	for_each_online_cpu(cpu) {
+		TfwRingbufferArea *rba = per_cpu_ptr(&rb_area, cpu);
+		atomic_set(&rba->unmapped, 1);
+		if (rba->pg) {
+			__free_pages(rba->pg, get_order(RBA_AREA_SIZE(rba)));
+			rba->pg = NULL;
+		}
+	}
+}
```

Comment:
What happens if we unload the module while the user space still didn't close and/or unmap the file?

Reply:
The memory remains available to user space.
I thought about it and wanted to notify the user space with a special event. But maybe it would be more correctly to handle this situation on the RB level. Will think.